File: cpp/questions/generateMatrix.cpp
#include <array>
#include <vector>

using namespace std;

class Solution {
private:
    static constexpr std::array<std::array<int, 2>, 4> ds = {{{0, 1}, {1, 0}, {0, -1}, {-1, 0}}};
public:
    vector<vector<int>> generateMatrix(int n) {
        std::vector<std::vector<int>> ans(n, std::vector<int>(n));
        int l = 0;
        int d = 0;
        int r = 0;
        int c = 0;
        for (int p=1; p<=n*n; ++p) {
            ans[r][c] = p;
            r += ds[d][0];
            c += ds[d][1];
            switch (d) {
                case 0:
                    if (c == n-l-1) {
                        d = 1;
                    }
                    break;
                case 1:
                    if (r == n-l-1) {
                        d = 2;
                    }
                    break;
                case 2:
                    if (c == l) {
                        d = 3;
                    }
                    break;
                case 3:
                    if (r == l) {
                        d = 0;
                        ++r;
                        ++c;
                        ++l;
                    }
                    break;
                default:
                    break;
            }
        }
        return ans;
    }
};

File: python/questions/generateMatrix.py
from typing import List

class Solution:
    ds = [[0, 1], [1, 0], [0, -1], [-1, 0]]

    def generateMatrix(self, n: int) -> List[List[int]]:
        ans = [[0]*n for _ in range(n)]
        d = 0
        r = 0
        c = 0
        for p in range(1, n*n+1):
            ans[r][c] = p
            nr = r + self.ds[d][0]
            nc = c + self.ds[d][1]
            if nr<0 or nc>=n or nr>=n or nc<0 or ans[nr][nc]:
                d = (d + 1) % 4
                r += self.ds[d][0]
                c += self.ds[d][1]
            else:
                r = nr
                c = nc
        return ans
